File: src/main/java/offer/Offer53_1.java
package offer;

/**
 * @author jingxinwu
 * @date 2021-11-24 1:04 上午
 */
public class Offer53_1 {

    public static void main(String[] args) {
        System.out.println(new Offer53_1().search(new int[]{1}, 1));
    }

    public int search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int l = 0;
        int r = nums.length - 1;
        int count = 0;
        while (l <= r) {
            int mid = (r - l) / 2 + l;
            if (nums[mid] == target) {
                count++;
                int temp = mid;
                while (temp > 0 && nums[temp - 1] == target) {
                    count++;
                    temp--;
                }
                int temp1 = mid;
                while (temp1 < nums.length - 1 && nums[temp1 + 1] == target) {
                    count++;
                    temp1++;
                }
                return count;
            } else if (nums[mid] < target) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }

        return count;
    }

}

File: go/src/leetcode/offer/Offer53_1.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 23:01
 */

func search(nums []int, target int) int {
	if len(nums) == 0 {
		return 0
	}
	left := searchL(nums, target)
	right := searchR(nums, target)
	if left == -1 || right == -1 {
		return 0
	}
	return right - left + 1
}

func searchL(nums []int, target int) int {
	idx := -1
	left, right := 0, len(nums)-1
	for left <= right {
		mid := (left + right) / 2
		if nums[mid] == target {
			idx = mid
		}
		if nums[mid] >= target {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return idx
}

func searchR(nums []int, target int) int {
	idx := -1
	left, right := 0, len(nums)-1
	for left <= right {
		mid := (left + right) / 2
		if nums[mid] == target {
			idx = mid
		}
		if nums[mid] <= target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return idx
}
